### DISCUSSION TITLE
Setting global floating number length

### DISCUSSION THREAD

**Discussion by lainisourgod** [Textualize/rich#1615] [Q&A]
Hi! I want to round all floating numbers like I do in `f"{number:0.3}"`. I can do this for bare numbers that I have in context, but if some object with numbers inside is pretty-printed, I can't change floats' formatting.
```python
>>> object_with_floats = {"a": 1.287777777778124, "b": [1.21242141244, -12.2124142214444444444]}
>>> rich.print(objects_with_floats)
{"a": 1.287777777778124, "b": [1.21242141244, -12.2124142214444444444]}
```
<img width="842" alt="image" src="https://user-images.githubusercontent.com/24773652/137997821-6aaaf992-3482-47b3-a439-f418a7e7ebea.png">

What I want
<img width="319" alt="image" src="https://user-images.githubusercontent.com/24773652/137998205-182a038d-ee4b-49bc-973f-83afba8b7e57.png">

**Comment by willmcgugan** [comment]
Hmm. Rich doesn't attempt to do more with the data than repr would do.

I'd consider it. Could you propose how you would like it to work?

**Reply by lainisourgod** [reply]
I'm glad you consider this to be a viable feature.

I failed to find how rich's repr works and where should this logic be put internally. I guess during tree traversal for every float we should call

```python
def format_float(number: float, pattern: str) -> str:
    """
    pattern is a format-pattern from https://pyformat.info/
    e.g. "{:.3f}"
    """
     return pattern.format(number)
```

Then we can add `pattern` argument to `log` and `print` calls
```python
console.log({1: {2: 123.4567890, 3: -35.7674}}, float_format_pattern="{:.3f}")
```

Also we can extend this functionality to format not only floating points but also regular numbers or strings via a `Mapping` argument:
```python
console.log({1: {2: 123.4567890, 3: -35.7674}}, format_patterns={float: "{:.3f}", str: "{:>10}"})
```

**Comment by zoeqevans** [comment]
I'd love to take a crack at implementing this, if it's still of interest to anyone?

Approximately all calls I make to the Rich print method look like `rprint(round(num, 3))` or some equivalent, so this would make my logging code a lot prettier. I can see few instances when you'd ever want to actually print a float in its full glory.
